### PR TITLE
Fix unhandled miner exception

### PR DESCRIPTION
**Pull request by dangershony**
Fix issue
```
Stratis.Bitcoin.Consensus.ConsensusException: input missing/spent
   at Stratis.Bitcoin.Consensus.ConsensusManager.BlockMinedAsync(Block block) in C:\Users\fullnodedan\Documents\Github\StratisBitcoinFullNode\src\Stratis.Bitcoin\Consensus\ConsensusManager.cs:line 281
   at Stratis.Bitcoin.Features.Miner.Staking.PosMinting.CheckStakeAsync(Block block, ChainedHeader chainTip) in C:\Users\fullnodedan\Documents\Github\StratisBitcoinFullNode\src\Stratis.Bitcoin.Features.Miner\Staking\PosMinting.cs:line 490
   at Stratis.Bitcoin.Features.Miner.Staking.PosMinting.GenerateBlocksAsync(WalletSecret walletSecret, CancellationToken cancellationToken) in C:\Users\fullnodedan\Documents\Github\StratisBitcoinFullNode\src\Stratis.Bitcoin.Features.Miner\Staking\PosMinting.cs:line 408
   at Stratis.Bitcoin.Features.Miner.Staking.PosMinting.<>c__DisplayClass38_0.<<Stake>b__0>d.MoveNext() in C:\Users\fullnodedan\Documents\Github\StratisBitcoinFullNode\src\Stratis.Bitcoin.Features.Miner\Staking\PosMinting.cs:line 288
[2018-12-11 18:44:18.7424 62] INFO: Stratis.Bitcoin.Utilities.AsyncLoop+<>c__DisplayClass16_0+<<StartAsync>b__0>d.MoveNext PosMining.Stake stopping.
[2018-12-11 18:44:18.7424 62] FATAL: Stratis.Bitcoin.Utilities.AsyncLoop+<>c__DisplayClass16_0+<<StartAsync>b__0>d.MoveNext PosMining.Stake threw an unhandled exception
[2018-12-11 18:44:18.7424 62] ERROR: Stratis.Bitcoin.Utilities.AsyncLoop+<>c__DisplayClass16_0+<<StartAsync>b__0>d.MoveNext PosMining.Stake threw an unhandled exception: Stratis.Bitcoin.Consensus.ConsensusException: input missing/spent
   at Stratis.Bitcoin.Consensus.ConsensusManager.BlockMinedAsync(Block block) in C:\Users\fullnodedan\Documents\Github\StratisBitcoinFullNode\src\Stratis.Bitcoin\Consensus\ConsensusManager.cs:line 281
   at Stratis.Bitcoin.Features.Miner.Staking.PosMinting.CheckStakeAsync(Block block, ChainedHeader chainTip) in C:\Users\fullnodedan\Documents\Github\StratisBitcoinFullNode\src\Stratis.Bitcoin.Features.Miner\Staking\PosMinting.cs:line 490
   at Stratis.Bitcoin.Features.Miner.Staking.PosMinting.GenerateBlocksAsync(WalletSecret walletSecret, CancellationToken cancellationToken) in C:\Users\fullnodedan\Documents\Github\StratisBitcoinFullNode\src\Stratis.Bitcoin.Features.Miner\Staking\PosMinting.cs:line 408
   at Stratis.Bitcoin.Features.Miner.Staking.PosMinting.<>c__DisplayClass38_0.<<Stake>b__0>d.MoveNext() in C:\Users\fullnodedan\Documents\Github\StratisBitcoinFullNode\src\Stratis.Bitcoin.Features.Miner\Staking\PosMinting.cs:line 288
--- End of stack trace from previous location where exception was thrown ---
   at Stratis.Bitcoin.Utilities.AsyncLoop.<>c__DisplayClass16_0.<<StartAsync>b__0>d.MoveNext() in C:\Users\fullnodedan\Documents\Github\StratisBitcoinFullNode\src\Stratis.Bitcoin\Utilities\AsyncLoop.cs:line 140
```

### CODE REVIEW

**Comment by noescape00** [comment]
@dangershony do you have full logs for this? 
I think we need to investigate under what circumstances miner produced a block with tx that lacks inputs, that sounds like a bug we need to fix instead of ignoring it.

File: src/Stratis.Bitcoin.Features.Miner/Staking/PosMinting.cs [hunk 1]
```diff
@@ -308,6 +308,14 @@ await this.GenerateBlocksAsync(walletSecret, this.stakeCancellationTokenSource.T
                     this.logger.LogDebug("Consensus error exception occurred in miner loop: {0}", cee.ToString());
                     this.rpcGetStakingInfoModel.Errors = cee.Message;
                 }
+                catch (ConsensusException ce)
+                {
+                    // All consensus exceptions should be ignored. It means that the miner
+                    // run into problems while constructing block or verifying it
+                    // but it should not halted the staking operation.
+                    this.logger.LogDebug("Consensus exception occurred in miner loop: {0}", ce.ToString());
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Features.Miner/Staking/PosMinting.cs]: Probably log on error level

**Comment by dangershony** [comment]
@noescape00 I suspect a race condition.

Block arrived to consensus and got validated and we set the tip, now the miner starts to mine on top of that new block ant takes trx from mempool, (it holds the lock of the mempool) and the block is now signalled to mempool to remove the trx that it had included already.

File: src/Stratis.Bitcoin.Features.Miner/Staking/PosMinting.cs (hunk 1 above)
Comment (dangershony) [reply src/Stratis.Bitcoin.Features.Miner/Staking/PosMinting.cs]: Its similar ot the other exception trace levels

**Pull request merged**